[menu] Overhaul the component API

michaldudak opened the pull request:
Closes #214
Closes #9
Closes #46
Closes #63
Closes #411

## Preview
https://deploy-preview-468--base-ui.netlify.app/base-ui/react-menu/

## To do
- [x] #411 
- [x] #46 
- [x] #63 
- [x] #9 
- [X] Nested menus
- [x] Animations
- [x] Docs
- [x] Configure <kbd>Esc</kbd> key behavior (close all/close one submenu level)
- [x] `keepMounted` prop
- [x] Configurable delay and open on hover

### In separate PRs
- [x] #52 
- [x] #53 
- [x] #70

michaldudak committed:
Move related components under Menu

michaldudak committed:
Rename Menu components and types

michaldudak committed:
useMenuRoot

michaldudak committed:
Move hooks under components

michaldudak committed:
Docs page

michaldudak committed:
Render props

michaldudak committed:
Fix types in tests

michaldudak committed:
Create a simple Positioner

michaldudak committed:
Overhaul menu reducers WIP

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu-on-hooks

michaldudak committed:
Refactor listReducer

michaldudak committed:
Remove unnecessary ListReducerAction type

michaldudak committed:
Refactor useList

michaldudak committed:
Clean up useControllableReducer

michaldudak committed:
Make useList tests pass

michaldudak committed:
Use new conformance tests

michaldudak committed:
Move menuReducer logic to dropdownReducer

michaldudak committed:
Few renames

michaldudak committed:
Refactor MenuRoot and MenuPopup

michaldudak committed:
Create MenuPopupContext

michaldudak committed:
Refactor useMenuItem

michaldudak committed:
Improve prop merging

michaldudak committed:
Fixes

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu-on-hooks

michaldudak committed:
Fix tests

michaldudak committed:
More tests

michaldudak committed:
Refactor Tabs to use new patterns

michaldudak committed:
Cleanup

michaldudak committed:
Remove value from list item metadata

michaldudak committed:
Fix tabs

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu-on-hooks

michaldudak committed:
Implement positioner

michaldudak committed:
Move more logic to useMenuItem hook

michaldudak committed:
Subitem trigger

michaldudak committed:
Nested menu using Floating UI

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu-on-hooks

michaldudak committed:
Nested menus

michaldudak committed:
Tests

michaldudak committed:
API docs

michaldudak committed:
Fix a few eslint errors

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

michaldudak committed:
Fix TS errors in legacy components

michaldudak committed:
Remove getItemState from useList

michaldudak committed:
useListNavigation

michaldudak committed:
typeahead

michaldudak committed:
Remove unneeded code

michaldudak committed:
Fix demo outlines

colmtuite commented:
@michaldudak 

- Submenu shouldn't close when the pointer leaves its popup. It should close when focus moves to a menuitem that is not the submenu's trigger.
- When you open a submenu via pointer, focus should remain on its trigger. Pressing right should move focus into the submenu.
- 750ms typeahead threshold is too long.
- Vid attached of placing pointer over a menuitem to focus it, then pressing up, and two menuitems are highlighted. Only one menuitem should ever be focused.
- Vid attached of some crazy ass bug

https://github.com/mui/base-ui/assets/805073/044b66aa-fa08-429c-87d8-51d18ee04f56
https://github.com/mui/base-ui/assets/805073/294742ba-0bcf-4034-abd8-5b5f318eac67

michaldudak committed:
Fix few issues

michaldudak commented:
@colmtuite 

Fixed:
- submenus don't close on mouseout anymore
- focus stays on the highlighted submenu trigger after it opens
- changed typeahead threshold to 350ms
- bugs you recorded are fixed

colmtuite commented:
@michaldudak When you close the menu via ESC, focus is not returned to the trigger.

atomiks commented on packages/mui-base/src/Menu/Root/useMenuRoot.ts:
```diff
@@ -192,8 +195,9 @@ export function useMenuRoot(parameters: UseMenuRootParameters) {
   const typeahead = useTypeahead(floatingRootContext, {
     listRef: itemLabels,
     activeIndex,
+    resetMs: 350,
```
@colmtuite  Why so short? For slow typers this seems way too fast to me. macOS' typeahead for Menus is over 500ms (not sure on the exact value, but 750ms seems like a good default).

colmtuite replied:
iirc, @michaldudak suggested 350ms? Not sure. I don't think I had a specific preference, just that the original value felt clearly too long.

atomiks commented on packages/mui-base/src/Menu/Root/menuReducer.ts:
```diff
@@ -0,0 +1,45 @@
+    case MenuActionTypes.open: {
+      const updateHighlight = action.event instanceof KeyboardEvent;
+
+      return {
+        ...state,
+        open: true,
+        highlightedValue: updateHighlight
+          ? moveHighlight(null, 'start', state.items, state.settings)
+          : state.highlightedValue,
+      };
```
What is the purpose of this reducer? `useListNavigation` handles the highlight (focus) all by itself, so this is likely conflicting and causing the possible bugs

michaldudak replied:
Without handling this here I had to press the right arrow twice to actually focus a first element in the menu.

michaldudak committed:
Declutter useMenuRoot types

michaldudak committed:
One click interaction

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

colmtuite commented:
@michaldudak Clicking the trigger repeatedly should toggle the menu open/closed

DiegoAndai commented:
Is the space/enter key opening broken? I remember it worked last week (coming here from the monthly meeting link 😅)


https://github.com/mui/base-ui/assets/16889233/06682b75-3472-4b5b-a71d-3935129bfc9e

michaldudak commented:
It is for now, but bear in mind it's still a draft and I'm working on it :)

michaldudak committed:
Fix Menu tests

michaldudak committed:
Fixing the tests

michaldudak committed:
Update internal-test-utils

michaldudak committed:
Remove react-test-renderer

michaldudak committed:
Remove enzyme

michaldudak committed:
Fix skipped tests

michaldudak committed:
Merge branch 'test-utils-update' into menu

michaldudak committed:
Fix listreducer tests

michaldudak committed:
Wrap render in act

michaldudak committed:
Merge branch 'test-utils-update' into menu

michaldudak committed:
Fix Menu tests

michaldudak committed:
Focus disabled items, orientation and direction support

michaldudak committed:
Fix type errors in slider

colmtuite commented:
When you move the cursor very quickly across menu items, it seems like focus is not able to keep up.
https://github.com/mui/base-ui/assets/805073/7f1f33cb-172f-47aa-9519-bc8f76493625

michaldudak committed:
Fix Slider

michaldudak committed:
Update MenuRoot tests

michaldudak commented:
> Clicking the trigger repeatedly should toggle the menu open/closed

In MacOS, the native menus open on the trigger's mousedown but close on mouseup. WDYT about adopting this behavior?

michaldudak committed:
Unmount the closed menu

michaldudak committed:
Close the menu on trigger click

colmtuite commented:
MacOS has some sort of interval. If you click, hold for 50ms, then release over the trigger, the menu remains open. If you click, hold for 500ms, then release over the trigger, the menu closes.

After chatting, we decided this might have two issues:

1. It may be problematic for some older or disabled users who can't confidently click the button fast enough.
2. There will be some perf hit to set a timer on each click.

So we decided to close the menu when you click, hold, drag, and release outside of the trigger.

michaldudak committed:
Use FloatingList

michaldudak committed:
Remove reducers infra and useCompound

michaldudak committed:
Fixing tests

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

michaldudak committed:
Remove unrelated changes

michaldudak committed:
Tests

michaldudak committed:
Unbind event handlers

michaldudak committed:
Fix click and drag

michaldudak committed:
Fix click and drag on submenus

michaldudak committed:
API docs

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

michaldudak committed:
Fix unit tests

michaldudak committed:
Fix karma tests

michaldudak committed:
Make tests more reliable

michaldudak committed:
Lint

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

michaldudak committed:
closeOnClick

michaldudak committed:
Animations

atomiks commented:
Some bugs:

1. The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest

https://github.com/user-attachments/assets/68f68436-15bb-48d9-a52a-4ba61896eb37

2. The submenu is left open when quickly navigating with ArrowLeft/ArrowRight even though focus is on the parent trigger

https://github.com/user-attachments/assets/9fe69fcc-f3e4-4353-9e05-aa11860c56f3

michaldudak committed:
Escape key behavior

michaldudak committed:
Docs

michaldudak committed:
Move types to namespaces

michaldudak commented:
> The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest

This looks like a bug in Floating UI. The same behavior can be seen in https://codesandbox.io/p/sandbox/admiring-lamport-5wt3yg.

I can take a look at it, unless you have an idea how to solve it easily.

> The submenu is left open when quickly navigating with ArrowLeft/ArrowRight even though focus is on the parent trigger

This one's interesting. I was only able to reproduce it when animations are enabled, and it seems like the first element of the submenu isn't focused when the menu is open. I'm still investigating the root cause.

michaldudak committed:
Fix disabling animations

atomiks commented:
> This looks like a bug in Floating UI. The same behavior can be seen in https://codesandbox.io/p/sandbox/admiring-lamport-5wt3yg.

I don't see the bug in that sandbox? When it rests the focus synchronizes. (I think it may be manually handled externally in the components, not the internal hooks though)

michaldudak commented:
What do you think about not disabling pointer events in these cases? This will also help in cases where someone wants to select a menu item below a submenu trigger right after the submenu opens.

michaldudak committed:
Do not block pointer-events in safe polygon

michaldudak committed:
Remove duplicated test

michaldudak committed:
JSDocs

michaldudak committed:
Fix tests

michaldudak committed:
Fix the docs build

michaldudak committed:
Fix the Tailwind demo

michaldudak committed:
Update Floating UI

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

atomiks commented:
- [Do not block pointer-events in safe polygon](https://github.com/mui/base-ui/pull/468/commits/15ffd83a65c771becd3e9d8ff8b0502d1da04d1c) - why was this added?
- If another button is focused instead of the parent menu trigger, and a submenu item is clicked, it returns focus to that instead of the parent. The top-level menu items aren't affected.

michaldudak committed:
keepMounted

michaldudak committed:
openOnHover

michaldudak commented:
> [Do not block pointer-events in safe polygon](https://github.com/mui/base-ui/pull/468/commits/15ffd83a65c771becd3e9d8ff8b0502d1da04d1c) - why was this added?

This is a workaround for "The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest".

> If another button is focused instead of the parent menu trigger, and a submenu item is clicked, it returns focus to that instead of the parent. The top-level menu items aren't affected.

Could you record a video showing this issue?

michaldudak committed:
Merge branch 'master' into menu

atomiks commented:
> Could you record a video showing this issue?

Bug seems to be gone in the latest deploy

> This is a workaround for "The focus doesn't synchronize once pointer-events stop being blocked and the cursor is at rest".

It only needs to synchronize focus for a submenu trigger, not a menu item. This is because the submenu item opens a menu on hover, so it needs to focus style applied. The Floating UI menu demo handles only that

michaldudak committed:
Fixes

michaldudak committed:
escapeClosesParents -> closeParentOnEscape

michaldudak committed:
Fix focus restoration when keepMounted=true

michaldudak committed:
Fixes to fixes

michaldudak committed:
Demo styles

michaldudak commented:
@atomiks I restored `blockPointerEvents` and managed to solve the issue with CSS.

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

michaldudak committed:
Fix flaky tests

michaldudak committed:
Merge branch 'master' into menu

michaldudak committed:
Merge branch 'master' into menu

michaldudak committed:
Fix subbomponents in trigger causing the menu to close immediately

michaldudak committed:
closeParentOnEscape -> closeParentOnEsc

atomiks commented:
Bug when traversing back to the submenu trigger, menu doesn't open

https://github.com/user-attachments/assets/162b1b95-74bb-4a82-9fc1-927a6b835d00

michaldudak committed:
Add MenuArrow

colmtuite commented:
@michaldudak Feedback based on our chat:

1. Add the Arrow component
2. Consider making focus looping configurable. I vote it should be.
3. Rename `positionStrategy` to `positionMethod` across all components.

michaldudak committed:
loop prop

michaldudak committed:
Arrow docs

michaldudak committed:
Update animation docs

michaldudak commented:
@colmtuite I did 1 and 2. `positionStrategy` rename should be done on the other components first in a separate PR.

michaldudak committed:
Fix submenu not reopening

michaldudak commented:
> Bug when traversing back to the submenu trigger, menu doesn't open

@atomiks, Should be working OK now.

> In this vid, the menu is initially opening above the trigger, then quickly repositioning itself below.

@colmtuite, could you prepare a reproduction on codesandbox?

oliviertassinari commented on packages/mui-base/src/Menu/Positioner/MenuPositionerContext.ts:
```diff
@@ -0,0 +1,29 @@
+
+export const MenuPositionerContext = React.createContext<MenuPositionerContext | null>(null);
+
+MenuPositionerContext.displayName = 'MenuPositionerContext';
```
```suggestion
if (process.env.NODE_ENV !== 'production') {
  MenuPositionerContext.displayName = 'MenuPositionerContext';
}
```

michaldudak committed:
Olivier's feedback

michaldudak committed:
data-state -> data-menu

michaldudak committed:
Merge remote-tracking branch 'upstream/master' into menu

michaldudak committed:
Fix types

michaldudak committed:
Merge branch 'master' into menu

atomiks reviewed (approved):
Looks great!
